src/main/java/org/elasticsearch/index/search/geo/GeoPolygonFilter.java

```diff
@@ -93,12 +93,24 @@ protected boolean matchDoc(int doc) {
 
         private static boolean pointInPolygon(GeoPoint[] points, double lat, double lon) {
             boolean inPoly = false;
+            double lat0, lon0, lat1, lon1;
+            if (lon < 0) {
+                lon += 360;
+            }
+
+            if (lat <0) {
+                lat += 180;
+            }
 
             for (int i = 1; i < points.length; i++) {
-                if (points[i].lon() < lon && points[i-1].lon() >= lon
-                        || points[i-1].lon() < lon && points[i].lon() >= lon) {
-                    if (points[i].lat() + (lon - points[i].lon()) /
-                            (points[i-1].lon() - points[i].lon()) * (points[i-1].lat() - points[i].lat()) < lat) {
+                lon0 = (points[i-1].lon() < 0) ? points[i-1].lon() + 360.0 : points[i-1].lon();
```

Comment:
Instead of reapplying the logic for the previous element on every iteration, you could initialize `lon0` and `lat0` to points[0] before the loop, and then add `lon0 = lon1; lat0 = lat1;` at the end of the loop?